snapcraft/commands/upload.py

```diff
@@ -107,6 +153,54 @@ def run(self, parsed_args):
         emit.message(message)
 
 
+def _validate_components(
+    provided_components: list[ComponentOption], snap_metadata: SnapMetadata
```

Comment:
It's pedantic and probably unnecessary, but `provided_components` could be a Collection or possibly even an Iterable.

```suggestion
    provided_components: abc.Collection[ComponentOption], snap_metadata: SnapMetadata
```